### PR TITLE
WT-6850 Creation of the workload generator class

### PR DESCRIPTION
**Pull request by etienneptl**

### CODE REVIEW

File: dist/test_data.py [hunk 1]
```diff
@@ -41,16 +41,38 @@ def __gt__(self, other):
...
+
+key_config=[
+    Config('key_size', 0, r'''
+        The size of the keys to be created''', min=0, max=10000),
+]
+value_config = [
+    Config('value_size', 0, r'''
+        The size of the values to be created''', min=0, max=10000),
+]
+scale_config = [
```
Comment (luke-pearson) [inline dist/test_data.py:53]: I think we should separate config's with newlines. Which would be consistent with `api_data.py`. So space out key_config, value_config, etc.

File: test/cppsuite/test_harness/api_const.h [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+
+static const char *CONNECTION_CREATE = "create";
+static const char *COLLECTION_COUNT = "collection_count";
+static const char *KEY_COUNT = "key_count";
+static const char *VALUE_SIZE = "value_size";
+
+} // namespace api_const
+} // namespace test_harness
+
+#endif
```
Comment (luke-pearson) [inline test/cppsuite/test_harness/api_const.h:17]: Style: Newline goes at end of file.

File: test/cppsuite/test_harness/random_generator.h [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+#define RANDOM_GENERATOR_H
+
+#include <algorithm>
+#include <cstddef>
+#include <random>
+#include <string>
+
+namespace test_harness {
+
+/* Generate random values. */
```
Comment (luke-pearson) [inline test/cppsuite/test_harness/random_generator.h:12]: This comment could be worded better, something like "Helper class to generate random values".

File: test/cppsuite/test_harness/random_generator.h [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+class random_generator {
+
```
Comment (luke-pearson) [inline test/cppsuite/test_harness/random_generator.h:14]: Don't need newline between class declaration and members.

File: test/cppsuite/test_harness/random_generator.h [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+    random_generator()
+    {
+        _generator = std::mt19937(_random_device());
+        distribution = std::uniform_int_distribution<>(0, CHARACTERS.size() - 1);
+    }
+
+    public:
+    static random_generator *
+    getInstance()
```
Comment (luke-pearson) [inline test/cppsuite/test_harness/random_generator.h:30]: get_instance()

File: test/cppsuite/test_harness/random_generator.h (hunk 4 above)
Comment (luke-pearson) [inline test/cppsuite/test_harness/random_generator.h:14]: Could you include a `private:` label, and as characters is a member of this class I think `_characters` would make more sense.

File: test/cppsuite/test_harness/workload_generator.h [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+#include <cstdint>
+#include <map>
+#include <vector>
+
+extern "C" {
+#include "test_util.h"
+}
+
+#include "api_const.h"
+#include "random_generator.h"
```
Comment (luke-pearson) [inline test/cppsuite/test_harness/workload_generator.h:14]: Lets be pedantic and do our includes in the following order:
 - Standard includes, alphabetically sorted
 - extern C includes, alphabetically sorted
 - local includes, alphabetically sorted

Can you fix up the includes in this files and other ones in the project, if they don't match this rule?

**Comment by luke-pearson** [comment]
@tammybailey, @quchenhao We're removing the namespace api_const and keeping the `const char *`'s for now. We'll leave the WT_RET and C++ify the configuration framework in a different ticket. We'll also create a ticket for improving the name of the `test_harness`.

File: test/cppsuite/test_harness/configuration_settings.h [hunk 7]
```diff
@@ -45,7 +42,7 @@ class configuration {
     std::string
     get_config()
     {
-        return _config;
+        return (_config);
```
Comment (quchenhao) [inline test/cppsuite/test_harness/configuration_settings.h:45]: So for the c++ coding style, I see we still return in brackets. Is this a decision made?

Comment (luke-pearson) [reply test/cppsuite/test_harness/configuration_settings.h:45]: This is a question I'll bring up in the Wednesday meeting, I think we'll have to have a wiki page defining the WiredTiger C++ style. Or opt to use the WiredTiger C style. For now lets just stick with the brackets until we've had that meeting. Currently our other C++ code, Workgen and Xray have varied style between them.

File: test/cppsuite/test_harness/workload_generator.h [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+
+        cursor = nullptr;
+        collection_count = key_count = value_size = 0;
+        collection_name = "";
+
+        /* Create the working dir. */
+        testutil_make_work_dir(home);
+
+        /* Open connection. */
+        WT_RET(wiredtiger_open(home, NULL, test_harness::CONNECTION_CREATE, &_conn));
```
Comment (quchenhao) [inline test/cppsuite/test_harness/workload_generator.h:71]: I don't think we should call wiredtiger open here in the workload generator. I think we should do that outside in the higher level to be able to open wiredtiger with different configs. At least can we pass the open config into this function?

I'm totally fine if this is a temporary solution.

Comment (luke-pearson) [reply test/cppsuite/test_harness/workload_generator.h:71]: Open config will likely be part of the configuration_options passed in, until we have a need to pass open config to `wiredtiger_open` I think we can keep it like this.

Comment (luke-pearson) [reply test/cppsuite/test_harness/workload_generator.h:71]: So yeah this is a temporary solution.

**Review by quchenhao — APPROVED** [review]
lgtm

File: test/cppsuite/test_harness/random_generator.h [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+    random_generator()
+    {
+        _generator = std::mt19937(_random_device());
+        distribution = std::uniform_int_distribution<>(0, _characters.size() - 1);
+    }
+
+    static random_generator *instance;
+    std::mt19937 _generator;
+    std::random_device _random_device;
+    std::uniform_int_distribution<> distribution;
```
Comment (luke-pearson) [inline test/cppsuite/test_harness/random_generator.h:47]: This should be _distribution.

**Review by luke-pearson — APPROVED** [review]
LGTM with one minor change.

**Review by tammybailey — COMMENTED** [review]
A few more general observations:

- There is no standard approach to error handling. Some methods throw exceptions, some use `WT_RET`, etc. It's definitely worth putting some thought into a common error handling routine.
- There is a fair amount of `strcmp` and `strlen` usage.

**Comment by luke-pearson** [comment]
> There is no standard approach to error handling. Some methods throw exceptions, some use WT_RET, etc. It's definitely worth putting some thought into a common error handling routine.
We'll aim to solve this in WT-7026 when we remove WT_RET. Is that okay?

**Review by tammybailey — APPROVED** [review]
LGTM - I'd argue that the arg parsing could be improved but I believe the plan is to move that bit to a common class anyway.

File: test/cppsuite/test_harness/api_const.h [hunk 10]
```diff
@@ -0,0 +1,15 @@
+/* Include guard. */
```
[the remark this answers is not in the document]
Comment (etienneptl) [reply test/cppsuite/test_harness/api_const.h:1]: @luke-pearson, should we get rid of them ? I agree that it is probably unnecessary. Can be addressed in WT-7206

File: test/cppsuite/test_harness/api_const.h [hunk 11]
```diff
@@ -0,0 +1,15 @@
...
+#ifndef API_CONST_H
+#define API_CONST_H
+
+/* Define all constants related to WiredTiger APIs and testing. */
```
[the remark this answers is not in the document]
Comment (etienneptl) [reply test/cppsuite/test_harness/api_const.h:5]: @tammybailey @luke-pearson  Is there a style you would prefer ? Or that is already being used that we should stick to ?

File: test/cppsuite/tests/poc.cxx [hunk 12]
```diff
@@ -1,35 +1,78 @@
...
+
+    // Parse args
+    // -C   : Configuration
+    // -t   : Trace level
+    for (int i = 1; i < argc; ++i) {
+        if (strcmp(argv[i], "-C") == 0) {
+            if (i + 1 < argc) {
+                cfg = argv[++i];
+            } else {
+                throw std::invalid_argument("No value given for option " + std::string(argv[i]));
```
[the remark this answers is not in the document]
Comment (etienneptl) [reply test/cppsuite/tests/poc.cxx:56]: There is no added value in throwing an exception here. Exceptions and errors handling will be fixed in WT-7206

**Comment by etienneptl** [comment]
I will merge this PR and address everything that has been discussed here in WT-7026.

**Pull request merged**